### PR TITLE
Refactoring adding bridge classes

### PR DESCRIPTION
**Pull request by AntoninThioux**
There is a lot of stuff I still want to change, but this pull request encompasses the main change of decoupling MovementResource and UserResource. Changes probably need to be made before this pull request is accepted.

## List of all (small) changes:
- [X] split tests and business logic
- [X] Add creation and subscriptions models
- [X] Add tests for the creation and subscription models
- [X] Add controllers for creation and subscriptions
- [X] Add tests for controller for creation and subscriptions
- [X] Add on_creation() and on_end_creation() + on_subscribtion() and on_subscription_event() should be implemented using set()
- [X]  Add listener to subscriptions remove and add users links to the network
- [X]  Add function to actually create a movement in the creation controller
- [X]  Add listener to creation to add a subscription
- [X]  Remove current_movements(self) from users and associated tests
- [X]  Change send single tests to use the new subscribe method
- [X]  Remove active_users(self) from movement model and associated tests
- [X]  Remove leaderless(self) from movement model and associated tests
- [x] Remove users from the movement model and associated tests
- [X]  Remove _subscribe() from movement controller and associated tests
- [X]  Remove subscribe() from movement controller and associated tests
- [X]  Remove remove_user_from_movement() from movement controller and associated tests
- [X]  Move possible followers() and possible_leaders() from helpers to leader and follower controllers also move associated tests
- [X]  Remove user_in_movement() from movement controller and associated tests
- [ ] Move find_last_signal() from helper to leader and move associated tests
- [X]  Remove get_subscriptions() in follower controller and associated tests

### CODE REVIEW

File: gridt/models/movement_user_relation.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+from datetime import datetime
+from sqlalchemy import Column, Integer, DateTime, ForeignKey, String
+from sqlalchemy.orm import relationship
+
+from gridt.db import Base
+
+class MovementUserRelation(Base):
```
Comment (JoepoFloepo) [inline gridt/models/movement_user_relation.py:7]: The difference between Movement user relation and movement user association isn't very clear

File: gridt/models/movement_user_relation.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        Args:
+            user (User, optional): The User related to the movement. Defaults to None.
+            movement (Movement, optional): The Movement related to the user. Defaults to None.
+        """
+        self.user = user
+        self.movement = movement
+        self.time_added = datetime.now()
+
+    def is_ended(self):
```
Comment (JoepoFloepo) [inline gridt/models/movement_user_relation.py:48]: has_ended() sounds a bit better

File: gridt/tests/models/test_creation.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+        with freeze_time('2022-12-27 03:36:00'):
+            creation.end()
+
+        self.assertEqual(creation.time_removed, datetime(2022,12,27,3,36,00))
+        self.assertTrue(creation.is_ended())
+
+
+    def test_repr(self):
+        user = User("abc123", "test@test.com", "password")
```
Comment (JoepoFloepo) [inline gridt/tests/models/test_creation.py:44]: I'm not sure about python but in java i like to put these hardcoded string as constants at the top of the class, because then you can reuse them.

File: gridt/controllers/creation.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        dict: json representation of the new subscription
+    """
+    with session_scope() as session:
+        user = load_user(user_id)
+        movement = load_movement(movement_id)
+        
+        creation = Creation(user, movement)
+        session.add(creation)
+    
+        #TODO: should have an emit listener here to actually create the movement
```
Comment (JoepoFloepo) [inline gridt/controllers/creation.py]: todo and i'm not sure if it was part of the things you still wanted to do

File: gridt/tests/controllers/test_subscription.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        json_subscription = remove_subscription(user_1_id, movement_id)
+        self.assertDictEqual(assert_json_user, json_subscription['user'])
+        self.assertDictEqual(assert_json_movement, json_subscription['movement'])
+        self.assertFalse(json_subscription['subscribed'])
+
+        # Test trying to remove none existing subscription
+        with self.assertRaises(E.SubscriptionNotFoundError):
+            _get_subscription(user_2_id, movement_id, self.session)
+
+        # TODO: Test that a message is emitted
```
Comment (JoepoFloepo) [inline gridt/tests/controllers/test_subscription.py]: todo

File: gridt/tests/controllers/test_subscription.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        self.assertEqual(subscription1.user_id, subscription2.user_id)
+        self.assertEqual(subscription1.movement_id, subscription2.movement_id)
+
+    def test_is_subscribed(self):
+        user = self.create_user()
+        movement = self.create_movement() 
+        self.session.commit()
+
+        user_id = user.id
+        movement_id = movement.id
```
Comment (JoepoFloepo) [inline gridt/tests/controllers/test_subscription.py:57]: I see that some code is run for a lot of the tests, is it not possible in python to have a setup function that runs before each tests? That way the code is a bit more reusable and shorter.

File: gridt/models/movement_user_relation.py (hunk 1 above)
Comment (AntoninThioux) [reply gridt/models/movement_user_relation.py:7]: I agree! I discussed with Yori about rename movement user association to UserUserMovementLink (because it is an actual link in the network). I thought about it more today maybe another name could be LeaderFollowerLink

File: gridt/models/movement_user_relation.py (hunk 2 above)
Comment (AntoninThioux) [reply gridt/models/movement_user_relation.py:48]: Good point. Although, I like using is for booleans in this case has makes more sense

File: gridt/tests/models/test_creation.py (hunk 3 above)
Comment (AntoninThioux) [reply gridt/tests/models/test_creation.py:44]: Hmm... ya makes sense

File: gridt/models/movement_user_relation.py (hunk 1 above)
Comment (AntoninThioux) [reply gridt/models/movement_user_relation.py:7]: But I think this is better done in a new pull request

File: gridt/tests/controllers/test_subscription.py (hunk 6 above)
Comment (AntoninThioux) [reply gridt/tests/controllers/test_subscription.py:57]: It is possible. I was probably a bit too lazy with tests

**Comment by AntoninThioux** [comment]
I started working on updating the pull request based on feedback and things I still wanted to finish.

Here is a complete list of changes I will make:

## List of todos before the pull request can be accepted
 [ ] Remove unused imports user.py
 [ ] Remove todo line 28 user.py
 [ ] Remove follower_associations in user.py
 [ ] Remove movements association_proxy() in user.py (line 47)
 [ ] Move send_email() to email_template.py
 [ ] Remove user_associations in movement.py (line 43)
 [ ] Remove user association_proxy in movement.py (line 49)
 [ ] Remove unused imports in movement.py
 [ ] Add type suggestions for user and movement to movement_relation.py (line 36)
 [ ] Remove optional parameters in movement_relation.py (line 36)
 [ ] Rename is_ended to has_ended on line 48 in movment_relation.py
 [ ] Remove new subscription event listeners it should import leader.py and follower.py to set the initial followers instead
 [ ] Remove unsubscribe event listeners it should use leader.py and follower.py to remove followers and leaders instead
 [ ] Remove event listener-related tests in test_subscription.py
 [ ] Isolate the tests of subscription not to add initial followers/leaders and remove all followers/leaders
 [ ] Add extended subscription JSON to subscription.py
 [ ] Remove unused imports subscription.py
 [ ] Move extended movement JSON from helper.py to movement.py (should call extended JSON in subscription.py)
 [ ] Move the extended movement JSON tests for test_helper.py to test_movement.py
 [ ] Move part of new_movement_by_user to movement.py
 [ ] Remove TODO in creation.py (line 107) now in movement.py
 [ ] Remove new creation listeners should call subscription instead
 [ ] Remove remove creation listeners
 [ ] Remove tests associated with creation listeners
 [ ] follower.py remove used imports
 [ ] Move leaders() from helpers.py to follower.py and rename it to get_follower_leaders()
 [ ] Move _find_last_signal() from helpers.py to leaders.py and rename to get_last_leader_signal()
 [ ] Move leaders() tests from test_helpers.py to test_follower.py
 [ ] Move _find_last_signal() tests
 [ ] Remove unused imports in helper.py
 [ ] Change possible_leader() function in leader.py to not user User and Movement objects instead use their ids
 [ ] Add initialization for tests to remove duplicate code

File: gridt/tests/models/test_creation.py (hunk 3 above)
Comment (Drvanon) [reply gridt/tests/models/test_creation.py:44]: Something I would do differently now looking back, is replacing the `unittest` library from python with [pytest](https://docs.pytest.org/en/7.2.x/), which is kind of the industry standard. In that case, you could create [fixtures](https://docs.pytest.org/en/7.2.x/explanation/fixtures.html#about-fixtures) to create the User. Having the strings inline there is not a problem, since you already abstracted the "User" entity creation to a different test. An equivalent way of dealing with this in unittest I guess, would be to have a separate non test method on the class where you would create the test user.

Comment (Drvanon) [reply gridt/tests/models/test_creation.py:44]: I would say that this also forces you to test in a way that is more sensible/readable.
